core/consensus/babe/impl/babe_synchronizer_impl.cpp

```diff
@@ -3,138 +3,803 @@
+      std::shared_ptr<network::Router> router,
+      std::shared_ptr<libp2p::basic::Scheduler> scheduler,
+      std::shared_ptr<crypto::Hasher> hasher)
+      : block_tree_(std::move(block_tree)),
+        block_executor_(std::move(block_executor)),
+        router_(std::move(router)),
+        scheduler_(std::move(scheduler)),
+        hasher_(std::move(hasher)) {
+    BOOST_ASSERT(block_tree_);
+    BOOST_ASSERT(block_executor_);
     BOOST_ASSERT(router_);
+    BOOST_ASSERT(scheduler_);
+    BOOST_ASSERT(hasher_);
+
+    BOOST_ASSERT(app_state_manager);
+    app_state_manager->atShutdown([this] { stops_ = true; });
   }
 
-  void BabeSynchronizerImpl::request(const primitives::BlockId &from,
-                                     const primitives::BlockHash &to,
-                                     const libp2p::peer::PeerId &peer_id,
-                                     const BlocksHandler &block_list_handler) {
-    std::string from_str = visit_in_place(
-        from,
-        [](const primitives::BlockHash &hash) { return hash.toHex(); },
-        [](primitives::BlockNumber number) { return std::to_string(number); });
-    logger_->info("Requesting blocks from {} to {}", from_str, to.toHex());
+  void BabeSynchronizerImpl::enqueue(
+      const primitives::BlockInfo &block_info,
+      const libp2p::peer::PeerId &peer_id,
+      BabeSynchronizer::SyncResultHandler &&handler) {
+    if (isInQueue(block_info.hash)) {
+      handler(block_info);
+      return;
+    }
 
-    static std::random_device rd{};
-    static std::uniform_int_distribution<primitives::BlocksRequestId> dis{};
-    network::BlocksRequest request{
-        dis(rd),
-        network::BlocksRequest::kBasicAttributes,
-        from,
-        to,
-        network::Direction::ASCENDING,
-        static_cast<uint32_t>(app_configuration_.maxBlocksInResponse())};
-
-    return pollClients(request, peer_id, block_list_handler);
-  }
+    const auto &last_finalized_block = block_tree_->getLastFinalized();
 
-  std::shared_ptr<network::SyncProtocolClient>
-  BabeSynchronizerImpl::selectNextClient(
-      std::unordered_set<std::shared_ptr<network::SyncProtocolClient>>
-          &polled_clients) const {
-    // we want to ask each client until we get the blocks we lack, but the
-    // sync_clients_ set can change between the requests, so we need to keep
-    // track of the clients we already asked
-    std::shared_ptr<network::SyncProtocolClient> next_client;
-    for (const auto &[it, client] : sync_clients_->clients()) {
-      if (polled_clients.find(client) == polled_clients.end()) {
-        next_client = client;
-        polled_clients.insert(client);
-        break;
-      }
+    auto best_block_res =
+        block_tree_->getBestContaining(last_finalized_block.hash, boost::none);
+    BOOST_ASSERT(best_block_res.has_value());
+    const auto &best_block = best_block_res.value();
+
+    if (block_info == best_block) {
```

Comment:
Could you please add short comments above such if blocks with returns in this method so it would be easier to track what's happening?

Reply:
Comment is added